mmf/models/visual_bert.py

```diff
@@ -545,26 +552,31 @@ def format_state_key(cls, key):
             .replace("bert.classifier", "model.classifier")
         )
 
-    def forward(self, sample_list):
+    def forward(self, sample_list: Dict[str, Tensor]) -> Dict[str, Tensor]:
+        if torch.jit.is_scripting():
+            assert (
+                "image_feature_0" in sample_list
+            ), "Key 'image_feature_0' is required in TorchScript model"
+
         sample_list = self.update_sample_list_based_on_head(sample_list)
         sample_list = self.add_custom_params(sample_list)
         sample_list = self.flatten_for_bert(sample_list)
+        sample_list = self.add_post_flatten_params(sample_list)
 
         output_dict = self.model(
-            sample_list.input_ids,
-            sample_list.input_mask,
-            sample_list.attention_mask,
-            sample_list.token_type_ids,
-            sample_list.visual_embeddings,
-            sample_list.position_embeddings_visual,
-            sample_list.visual_embeddings_type,
-            sample_list.image_text_alignment,
-            sample_list.masked_lm_labels,
+            sample_list["input_ids"],
+            sample_list["input_mask"],
+            sample_list["attention_mask"],
+            sample_list["token_type_ids"],
+            sample_list["visual_embeddings"],
+            sample_list["visual_embeddings_type"],
+            getattr_torchscriptable(sample_list, "image_text_alignment", None),
+            getattr_torchscriptable(sample_list, "masked_lm_labels", None),
         )
 
-        if "pretraining" in self.config.training_head_type:
+        if self.training_head_type == "pretraining" and not torch.jit.is_scripting():
             loss_key = "{}/{}".format(
-                sample_list.dataset_name, sample_list.dataset_type
+                sample_list["dataset_name"], sample_list["dataset_type"]
             )
             output_dict["losses"] = {}
             output_dict["losses"][loss_key + "/masked_lm_loss"] = output_dict.pop(
```

Comment:
I think we should throw an error if `torch.jit.is_scripting() and self.training_head_type == "pretraining"`. Reason is we are not updating the loss in the output dict properly here along with dataset name and type, which might cause issues during multi task pretraining. So basically we won't support scripting mode during pretraining.

Reply:
Makes sense. I actually thought that there is nothing stopping you from actually doing pretraining in scripted mode except dataset_name and dataset_type so I kept the support for this open. This will require modifications to tests as well. I will make this change and update the diff.